### DISCUSSION TITLE
USD (米ドル) 豆知識

### DISCUSSION THREAD

**Discussion by murnana** [code-polaris/reading-circle-tdd#43] [General]
この本では `Doller.amount` が `int`…つまり整数で定義されていますが、
メンバ変数1つで表すのであれば、実数のほうが良いです。
というのは、米ドルの世界では「ドル」より小さい「セント」がいるからです。

100セント = 1ドル なので
1セント = 1/100 ドル = 0.01ドル となります。

「ドル」と「セント」をメンバ変数でそれぞれ分けたほうがいいか？はわからない 😅

----
さらに豆知識ですが、iOS/Androidで課金系となると、

- 値段表示用の文字列 (`○○円`、`＄○○.○)`
- ISO 4217 … 通貨の種類を表す文字列 (`JPY`, `USD`...)
- 通貨自身を表す実数 (`100.00`, `0.71`)

を渡してきます。

**Comment by shino51** [comment]
私の会社は外貨為替系なのですが、お金はすべて `long` 型でDBへの保存などもされています。これは、お金の最小単位をドルではなく、その通貨で使われる一番小さい数字に合わせているからです。
これは、1.00などと表わすと、通貨の計算の時に不確実な数字をたたき出してしまうことを避ける為だと理解しています。

つまり、コード上でUSDが`100`と表わされた場合、それは100セントの事であり、1ドルを表します。

うちの場合、Currencyというクラスがあって、そこに各通貨の情報が詰め込まれています。各通貨にはそれぞれfractionDigits という属性があり、それによってどこからがドルなのかどこまでがセントなのか、という事を表しています。(例えば、USDの時のfractionDigitsは2です)

多分、この本は分かりやすくする為にその辺をはしょっているんだと思いますけど、普段お金の管理をしているとちょっとモヤりますよね。
私だったらせめてBigDecimalを使うかなぁとは思います。笑

**Comment by rinahira** [comment]
普段日本円しか扱っていなくて基本小数点以下って考えたことなかったので
(使うとすれば消費税計算の切り捨て誤差対策くらい？セブンイレブンとかは小数点以下2桁で税込み表示してますよねー)
外貨を扱う際には気を付けないとなーと思いました！勉強になります～:pray:
